docs/features/interceptors.md

```diff
@@ -154,14 +158,34 @@ static class Program
 {
     public static void M<T2>(T2 t)
     {
-        C.InterceptableMethod(t);
+        C.InterceptableMethod(t); // intercepts with 'Interceptor1<T2>(T2)'
     }
 }
 
 static class D
 {
     [InterceptsLocation("Program.cs", 13, 11)]
-    public static void Interceptor1(object s) => throw null!;
+    public static void Interceptor1<T3>(T3 t3) => throw null!;
+}
+```
+
+If an interceptable method signature uses type parameters from the containing type (in parameters and returns, including `this`), it won't be possible to declare an interceptor for it in another type. This is because there's no position where the original containing type's type parameters can be substituted in to the interceptor. We could revisit this limitation if needed, perhaps by reviewing the arity matching requirements proposed for *implicit and explicit extensions* as a starting point.
```

Comment:
We need to resolve this issue to fix https://github.com/dotnet/aspnetcore/issues/47338. The methods we're trying to intercept in ASP.NET Core were never generic to begin with. We want the interceptor method to capture the generic parameter from an outer scope that was never passed to the original intercepted method, so the arity will be different.

Take the examples I originally posted in the ASP.NET Core issue:

```csharp
public static void TestGenericParam<TUser>(IEndpointRouteBuilder app) where TUser : class
{
    app.MapPost("/test-param", ([FromServices] UserManager<TUser> userManager) => { });
}

public static void TestGenericResponse<TUser>(IEndpointRouteBuilder app) where TUser : new()
{
    app.MapPost("/test-return", () => new TUser());
}
```

The original `MapPost` has a generic arity of zero. The goal was to be able to define a `MapPost<TUser>(IEndpointRouteBuilder app, string pattern, Delegate handler)` so we could then use `TUser` in the generated code to resolve the arguments and serialize the response correctly.

Of course, for `TUser` to be inferred at these call sites, `MapPost<TUser>` would need to take an `Action<UserManager<TUser>>` or `Func<TUser>` respectively, rather than a `Delegate` like it does in the current PR to make ASP.NET Core use this feature at https://github.com/dotnet/aspnetcore/pull/48555.

Looking at the signature matching part of this document, it appears arguments have to match the original method arguments exactly rather than take a more specific type. It'd be nice if we could relax that if we could statically prove that the call site will always use a more specific type that the interceptor accepts.

Here's an example that demonstrates the need for the interceptor to change the parameter type from `Delegate` to `Func<T2>` for type inference to work.

```csharp
class C
{
    [Interceptable]
    public static void InterceptableMethod(Delegate handler) => throw null!;
}

static class Program
{
    public static void M<T1>(T1 t)
    {
        C.InterceptableMethod(() => t); // intercepts with 'Interceptor<T1>(Func<T1>)'
    }
}

static class D
{
    [InterceptsLocation("Program.cs", 13, 11)]
    public static void Interceptor<T2>(Func<T2> handler) => throw null!;
}
```

Reply:
Thanks, this is good information and it's a tricky scenario. Is it safe to assume that we can't know the full set of possible type arguments to `TestGenericResponse<TUser>` in advance? (letting us possibly test and branch on the type of the return value.)

Reply:
Unfortunately, it's not safe to assume we can discover the full set of type arguments at compile time. I discovered this issue working on [`MapIdentityApi<TUser>()`](https://github.com/dotnet/aspnetcore/issues/47227) where the `TUser` is provided by the consumer of the library.